todoman/widgets.py

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2016 Hugo Osvaldo Barrera
+# Copyright (c) 2013-2016 Christian Geier et al.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import re
+
+import urwid
+
+
+class ExtendedEdit(urwid.Edit):
+    """A text editing widget supporting some more editing commands"""
+    def keypress(self, size, key):
+        if key == 'ctrl w':
+            self._delete_word()
+        elif key == 'ctrl u':
+            self._delete_till_beginning_of_line()
+        elif key == 'ctrl k':
+            self._delete_till_end_of_line()
+        elif key == 'ctrl a':
+            self._goto_beginning_of_line()
+        elif key == 'ctrl e':
+            self._goto_end_of_line()
+        elif key == 'ctrl d':
+            self._delete_forward_letter()
+        # TODO: alt b, alt f
+        else:
+            return super(ExtendedEdit, self).keypress(size, key)
+
+    def _delete_forward_letter(self):
+        text = self.get_edit_text()
+        pos = self.edit_pos
+        text = text[:pos] + text[pos + 1:]
+        self.set_edit_text(text)
+
+    def _delete_word(self):
+        """delete word before cursor"""
+        text = self.get_edit_text()
+        t = text[:self.edit_pos]
+
+        words = re.findall(r'[\w]+|[^\w\s]', t, re.UNICODE)
+        t = t.rstrip()
+        if t == '':
+            f_text = t
+        else:
+            f_text = t[:len(t) - len(words[-1])]
+
+        self.set_edit_text(f_text + text[self.edit_pos:])
+        self.set_edit_pos(len(f_text))
+
+    def _delete_till_beginning_of_line(self):
+        """delete till start of line before cursor"""
+        text = self.get_edit_text()
+        t = text[:self.edit_pos]
```

Comment:
I don't think this is necessary, you should be able to use the `start` and `end` parameters of `find` to avoid needless string slicing. (this is more of a performance improvement, not really for prettyness)